File: solutions/haskell/pythagorean-triplet/1/src/Triplet.hs
module Triplet (tripletsWithSum) where

tripletsWithSum :: Int -> [(Int, Int, Int)]
tripletsWithSum s =  [ (a, b, c) |
    c <- [1..s-2],
    b <- [1..c-1],
    let a = s-c-b,
    a > 0 && a < b,
    a*a + b*b == c*c ]

File: solutions/elixir/pythagorean-triplet/1/lib/triplet.ex
# For Shree DR.MDD

defmodule Triplet do
  @type triplet :: list(non_neg_integer)

  @spec sum(triplet) :: non_neg_integer
  def sum([x, y, z]), do: x + y + z
  def sum(_), do: :wrong_triplet

  @spec product(triplet) :: non_neg_integer
  def product([x, y, z]), do: x * y * z
  def product(_), do: :wrong_triplet

  def generate(total) when total > 0, do: generate(1, total, total)
  def generate(_), do: :error

  @spec pythagorean?(triplet) :: boolean
  def pythagorean?([x, y, z]), do: x * x + y * y == z * z
  def pythagorean?(_), do: :wrong_triplet

  defp generate(start, limit, total) do
    for i <- start..div(total, 3),
        j <- (i + 1)..(limit - 1),
        k = total - i - j,
        k > j,
        Triplet.pythagorean?([i, j, k]) do
      [i, j, k]
    end
  end
end
